### PR TITLE
 API Remove Packagist API and use Composer to determine recommended andd latest updates

### LINKED ISSUE #23: Use composer/semver for constraint comparison
To reduce maintenance burden, we could use the [composer/semver](https://github.com/composer/semver/) package to perform constraint matching/comparisons when looking for the latest version.

This would also be useful in determining the "next safe version" (https://github.com/bringyourownideas/silverstripe-maintenance/issues/33) which requires semver comparisons against the installed constraint - for example:

| Constraint | Candidate | Is safe |
| --- | --- | --- |
| ^1.0 | 1.0.0 | true |
| ^1.0 | 1.0.5 | true |
| ^1.0 | 1.1.5 | true |
| ^1.0 | 2.0.0 | false |
| ^1.0 | 3.2.1 | false |
| ~1.0.3 | 1.0.0 | false |
| ~1.0.3 | 1.0.3 | true |
| ~1.0.3 | 1.0.4 | true |
| ~1.0.3 | 1.0.9 | true |
| ~1.0.3 | 1.1.0 | false |
| ~1.0.3 | 2.0.0 | false |
| >=3.2 | 3.2.0 | true |
| >=3.2 | 3.3.4 | true |
| >=3.2 | 4.2.0 | true |

Note that the last constraint is a dangerous (unbounded) one, but using composer/semver to check these constraints will remove similar logic from this module and reduce our maintenance overhead.

### LINKED ISSUE #24: Determine "next safe version" for the current constraint
Related: #23 

Given a constraint of `~1.2.3` and a list of versions which contains `1.2.7`, `1.3.5` and `2.0.0`, the "next safe version" would be `1.2.7`.

### LINKED ISSUE #33: List module version
As a Product Owner
I want to know what version of each module
So that I can have a more informed discussion with my developers about upgrade decisions

## Acceptance criteria

- [x] The currently installed version of each module used is listed
- [x] The most current version of the module is listed 
- [x] Next 'safe' stable version available is listed

Designs: https://projects.invisionapp.com/share/ASGU6A7RT8U#/screens/291050684

## Tasks

- [x] Migrate silverstripe-composer-verisons model into silverstripe-maintenance module: https://github.com/bringyourownideas/silverstripe-maintenance/issues/46
- [x] Convert composer-update-checker module's model to an extension: https://github.com/bringyourownideas/silverstripe-composer-update-checker/issues/25
- [ ] Replace constraint checking with composer/semver: https://github.com/bringyourownideas/silverstripe-composer-update-checker/issues/23
- [ ] Determine next safe version: https://github.com/bringyourownideas/silverstripe-composer-update-checker/issues/24

### PR DESCRIPTION
**Pull request by robbieaverill**
High level changes:

* Removes Packagist API
* Adds Composer API (https://github.com/bringyourownideas/silverstripe-composer-update-checker/issues/23)
* Uses Composer for all package, dependency, constraint and version manipulation handling (removes a bit of the BuildTask)
* Adds Composer API access to the ComposerLoader in the main maintenance module
* Adds "Available" (semver safe upgrade) and "Latest" (latest) columns to the main report (https://github.com/bringyourownideas/silverstripe-composer-update-checker/issues/24)
* Moves some logic out of the BuildTask into UpdateChecker as a separate service class for easier mocking in tests

Resolves https://github.com/bringyourownideas/silverstripe-composer-update-checker/issues/23, resolves https://github.com/bringyourownideas/silverstripe-composer-update-checker/issues/24
 and closes https://github.com/bringyourownideas/silverstripe-maintenance/issues/33

### CODE REVIEW

**Comment by spekulatius** [comment]
Hello @robbieaverill 

thank you for working on this. I cloned the repo and had a look through your work. Looks all very good. The only thing I noticed were two missing doc blocks for private static arrays. Not sure if that is intentional - because it were the only two. Its in ComposerUpdateExtension and CheckComposerUpdatesTask, in case you want to add them. 

Cheers,
Peter

**Comment by robbieaverill** [comment]
Thanks for the review! Ah yep, they were intentionally omitted as they get inherited this way from DataObject, but I’m happy to add them anyway if you’d prefer :-)

**Comment by spekulatius** [comment]
Na, its fine. I was just wondering. Now I know :)

File: src/Tasks/CheckComposerUpdatesTask.php [hunk 1]
```diff
@@ -24,381 +30,108 @@ class CheckComposerUpdatesTask extends BuildTask
...
-                if ($result !== false) {
-                    $this->recordUpdate($package, $currentVersion, $result);
-                }
-            }
-        }
+        $allowedTypes = Config::inst()->get(UpdatePackageInfo::class, 'allowed_types');
 
-        // finished message
-        $this->message('The task finished running. You can find the updated information in the database now.');
+        return in_array($type, $allowedTypes);
```
Comment (NightJar) [inline src/Tasks/CheckComposerUpdatesTask.php]: Could this be cast to an array to catch the edge case where someone may set e.g. `allowed_types: silverstripe-vendormodule`?

File: src/Extensions/ComposerUpdateExtension.php [hunk 2]
```diff
@@ -21,21 +20,19 @@ class ComposerUpdateExtension extends DataExtension
...
-        'LatestVersion' => 'Varchar',
-    ];
-
-    private static $summary_fields = [
-        'LatestVersion',
+        'VersionConstraint' => 'Varchar(50)',
+        'AvailableVersion' => 'Varchar(50)',
+        'AvailableHash' => 'Varchar(50)',
+        'LatestVersion' => 'Varchar(50)',
+        'LatestHash' => 'Varchar(50)',
```
Comment (NightJar) [inline src/Extensions/ComposerUpdateExtension.php:27]: Defining some of these in a `private static $summary_fields` would be good, as a selection you wish to expose to CSV exported reports.

**Review by NightJar — APPROVED** [review]
2 minor comments, but looks really good :)
Nice work!

File: src/Extensions/ComposerUpdateExtension.php [hunk 3]
```diff
@@ -21,21 +20,19 @@ class ComposerUpdateExtension extends DataExtension
...
 
     /**
      * Automatically schedule a self update job on dev/build
      */
     public function requireDefaultRecords()
     {
-        Injector::inst()
-            ->get(QueuedJobService::class)
-            ->queueJob(new CheckComposerUpdatesJob());
+        Injector::inst()->get(QueuedJobService::class)->queueJob($this->getJobName());
```
Comment (NightJar) [inline src/Extensions/ComposerUpdateExtension.php]: Oh, actually :s

```
[Recoverable Error] Argument 1 passed to QueuedJobService::queueJob() must be an instance of QueuedJob, string given
```
`silverstripe-composer-update-checker\src\Extensions\ComposerUpdateExtension.php on line 35`

queuedjobs `2.10.5`

**Review by NightJar — CHANGES_REQUESTED** [review]
The queued job instantiation requires an instance, not a string in the resolved version between this module and `silverstripe-maintenance`.

**Comment by robbieaverill** [comment]
Updated. The test failures are unrelated and are caused by queuedjobs (although unrelated to that module too) - issue is here: https://github.com/silverstripe/silverstripe-framework/pull/8090

**Comment by NightJar** [comment]
Thanks for explaining :)

**Pull request merged**